client/trino-client/src/main/java/io/trino/client/MaterializingInputStream.java

```diff
@@ -51,26 +52,33 @@ public int read()
     }
 
     @Override
-    public int read(char[] cbuf, int off, int len)
+    public int read(byte[] buffer, int off, int len)
             throws IOException
     {
-        int read = super.read(cbuf, off, len);
+        int read = super.read(buffer, off, len);
         if (read > 0) {
-            int copyLength = Math.min(read, headChars.length - currentOffset);
+            int copyLength = Math.min(read, head.length - currentOffset);
             if (read > copyLength) {
                 remaining += read - copyLength;
             }
             if (copyLength > 0) {
-                System.arraycopy(cbuf, off, headChars, currentOffset, copyLength);
+                System.arraycopy(buffer, off, head, currentOffset, copyLength);
                 currentOffset += copyLength;
             }
         }
         return read;
     }
 
+    @Override
+    public int read(byte[] buffer)
+            throws IOException
+    {
+        return read(buffer, 0, buffer.length);
+    }
+
     public String getHeadString()
     {
-        return String.valueOf(headChars, 0, currentOffset) + (remaining > 0 ? format("... [" + bytesOmitted(remaining) + "]", remaining) : "");
+        return new String(head, 0, currentOffset, UTF_8) + (remaining > 0 ? format("... [" + bytesOmitted(remaining) + "]", remaining) : "");
```

Comment:
what will happen if you cut utf char in half here?

Reply:
probably we will display some garbage but I have no idea whether this matters

Reply:
IDK - I would reather expect an exception. Which is not what we want I guess

Reply:
I've added a graceful handling of the malformed buffer